compiler/vm/vmdef.nim

```diff
@@ -90,27 +91,224 @@ type
                       # temporary slot usage. This is required for the parameter
                       # passing implementation.
-    slotEmpty,        # slot is unused
-    slotFixedVar,     # slot is used for a fixed var/result (requires copy then)
-    slotFixedLet,     # slot is used for a fixed param/let
-    slotTempUnknown,  # slot but type unknown (argument of proc call)
-    slotTempInt,      # some temporary int
-    slotTempFloat,    # some temporary float
-    slotTempStr,      # some temporary string
-    slotTempComplex,  # some complex temporary (s.node field is used)
-    slotTempPerm      # slot is temporary but permanent (hack)
-
-
-  TFullReg* = object  # with a custom mark proc, we could use the same
-                      # data representation as LuaJit (tagged NaNs).
+    slotEmpty,        ## slot is unused
+    slotFixedVar,     ## slot is used for a fixed var/result (requires copy then)
+    slotFixedLet,     ## slot is used for a fixed param/let
+    slotTempUnknown,  ## slot but type unknown (argument of proc call)
+    slotTempInt,      ## some temporary int
+    slotTempFloat,    ## some temporary float
+    slotTempStr,      ## some temporary string
+    slotTempComplex,  ## some complex temporary (s.node field is used)
+    slotTempPerm      ## slot is temporary but permanent (hack)
+
+  AtomKind* = enum
+    # Trivial values
+    akInt
+    akFloat
+    akPtr
+
+    # Managed values
+    akSet
+    akString
+    akSeq
+    akRef
+    akCallable # XXX: rename to akProcedural or akFuncHandle
+
+    akDiscriminator
+
+    akPNode # XXX: This is two things, typedesc values in the context
+            # of type evaluation and `NimNode`
+            # Since these are somewhat different, and `typedesc` can't be stored in locations,
+            # it might make sense to introduce a new register kind for typedescs...
+
+    # Pseudo atom kinds
+    akObject
+    akArray
+
+
+  # XXX: `ref` until I decide on a better way to manage these
+  PVmType* = ref VmType
+
+  FieldPosition* = uint32
+
+  BranchListEntryKind* = enum
+    blekStart
+    blekBranch
+    blekEnd
+
+  BranchListEntry* = object
+    case kind*: BranchListEntryKind
+    of blekStart:
+      field*: FieldPosition
+      numItems*: uint32
+      numBranches*: uint16
+    of blekBranch:
+      fieldRange*: Slice[FieldPosition]
+    of blekEnd:
+      parent*: uint32
+      nextField*: FieldPosition
+
+  VmType* = object ## A `VmType` describes the type of a location
+    case kind*: AtomKind
+    of akInt, akFloat, akString, akPNode: discard
+    of akPtr, akRef:
+      targetType*: PVmType
+    of akSet:
+      discard
+    of akSeq:
+      seqElemStride*: int
+      seqElemType*: PVmType
+    of akCallable:
+      retType*: PVmType
+    of akDiscriminator:
+      # A discriminator consists of two things: the value as seen by the guest
+      # and the index of the branch
+      numBits*: int ## The amount of bits the discriminator value occupies
```

Comment:
A discriminator is special in order to prevent the guest from overwriting it (via a pointer). It actually stores two values: the value the user sees/uses and an internal branch number.

As as storage optimization, these two are compacted into single int value. If a discriminant has values in the range 0..12, `numBits` would be 4. The discriminator's value is stored in the lower 4 bits then and the branch number in the 4 bits after, only requiring one byte for storage.

The branch number denotes the active branch and is used when traversing the active fields (via `variantFields`) and is stored as part of the discriminator in order to get around having to store `discriminator value -> branch number` mappings for variant types.

When performing code generation for discriminator setup or changing, `vmgen` generates code to map the value to the branch number (`genDiscrVal` does this).

Once `vmgen` learns to generate code for copying and resetting objects, the whole variant object handling can be simplified a bit